### ISSUE TITLE
Add keyword to put space between bars/boxes/violins in groups

### ISSUE THREAD

**Issue by kescobo** [JuliaPlots/StatPlots.jl#194]
As discussed in #192:

Default:

![image](https://user-images.githubusercontent.com/3502975/47303266-787b9600-d5f1-11e8-84cb-7da6b4395f11.png)

Would like to give users the option to put a space between:

![image](https://user-images.githubusercontent.com/3502975/47303315-9517ce00-d5f1-11e8-8437-6faac8c49031.png)

**Comment by piever** [comment]
I'm facing the same issue in StatsMakie, any idea what a good keyword would be? I'm thinking `shrink  = 0.8` but maybe there are better options.

**Comment by mkborregaard** [comment]
R uses `space`

**Comment by daschw** [comment]
Does it make sense to use `barwidth` for this?

**Comment by mkborregaard** [comment]
Possibly but is the bar width and the spacing not two slightly different things you'd like independent control of?

**Comment by daschw** [comment]
I don't think they can be controlled seperately within one series. A bigger barwidth implies smaller spacing and vice versa, no?

**Comment by piever** [comment]
> I don't think they can be controlled seperately within one series. A bigger barwidth implies smaller spacing and vice versa, no?

In practice for a single series only one quantity is relevant: `space` is only effective when no explicit width is given and I think it makes a difference in terms of API. For example, let's say I do:

```julia
bar(x, y)
```

I would get a barwidth of `0.8*minimum(diff(x))`. If I want to increase that value systematically across plots, I could do `width = 0.9*minimum(diff(x))` but that's a bit clumsy, so maybe it'd be easier to have:

```julia
space --> 0.2
width --> (1-space)*minimum(diff(x))
```

And the user can manipulate space if needed (`space = 0` for example)

> R uses `space`

Is it multiplicative or additive? Meaning, would it be something like the above `width --> (1-space)*minimum(diff(x))` or more of a `width --> minimum(diff(x)) - space` ? I'd prefer the first one but was curious to know what's the default in other environments.

**Comment by mkborregaard** [comment]
In R it's additive. So it shifts the position of the center of the bar on the x axis by bar-width + space for each new bar.

**Comment by kescobo** [comment]
>I don't think they can be controlled separately within one series. A bigger barwidth implies smaller spacing and vice versa, no?

Not really...

<img width="552" alt="screen shot 2018-11-19 at 09 30 11" src="https://user-images.githubusercontent.com/3502975/48713235-c62df300-ebdd-11e8-8a83-cef6798401c6.png">

**Comment by mkborregaard** [comment]
Yeah there are a lot of factors - space-within-group , space-among-groups, bar-width and distance-between-group-centers (the x ticks). Different plotting packages and languages have different implementations. Really not sure what I find most intuitive, though I'm quite sure it is not `shrink` :-p

**Comment by kescobo** [comment]
>though I'm quite sure it is not shrink

Agree! 

>Yeah there are a lot of factors - space-within-group , space-among-groups, bar-width and distance-between-group-centers (the x ticks)

Yeah, it seems complex, but I actually think it only really requires additional control over 2 things - bar width and space-within-group (or space-among-groups, I'm not entirely sure what you mean with that - I'm referring to the space between the blue and purple bars in my example). The x values determines distance between group centers (it seems sensible to default to integers for categorical variables, but allow overrides).

That is, assuming we're expecting it all to be uniform across groups. If not, then it will definitely get much more complicated... I can imagine when one might want to change that, but I'm not sure I've seen it in the wild...

**Comment by Paethon** [comment]
Any progress on this?
